src/data/nat/nth.lean

```diff
@@ -40,367 +41,335 @@ variable (p : ℕ → Prop)
 /-- Find the `n`-th natural number satisfying `p` (indexed from `0`, so `nth p 0` is the first
 natural number satisfying `p`), or `0` if there is no such number. See also
 `subtype.order_iso_of_nat` for the order isomorphism with ℕ when `p` is infinitely often true. -/
-noncomputable def nth : ℕ → ℕ
-| n := Inf { i : ℕ | p i ∧ ∀ k < n, nth k < i }
+noncomputable def nth (p : ℕ → Prop) (n : ℕ) : ℕ :=
+by classical; exact
+  if h : set.finite (set_of p) then (h.to_finset.sort (≤)).nthd n 0
+  else @nat.subtype.order_iso_of_nat (set_of p) (set.infinite.to_subtype h) n
```

Comment:
The golfing looks amazing, but can you give us an estimate of how much is due to changing this definition vs simply breaking up long proofs into smaller lemmas?

Reply:
I don't remember. With the new definition, it's easier to prove lemmas because we already know properties of `list.nthd` and `nat.subtype.order_iso_of_nat`.